File: Array/zhanchengguo/0622-嵌套数组/Main.java
import android.util.Log;

public class Main {

    void main() {
        int a[] = {5,4,0,3,1,6,2};
        int result = arrayNesting(a);
        Log.w("result is ", result + "");
    }

    public int arrayNesting(int[] nums) {
        boolean[] visited = new boolean[nums.length];
        int res = 0;
        for (int i = 0; i < nums.length; i++) {
            if (!visited[i]) {
                int start = nums[i], count = 0;
                do {
                    start = nums[start];
                    count++;
                    visited[start] = true;
                }
                while (start != nums[i]);
                res = Math.max(res, count);
            }
        }
        return res;
    }
}

File: Array/zhanchengguo/0622-嵌套数组/main.kt

fun main() {
    val s = arrayOf(5, 4, 0, 3, 1, 6, 2)
    val result = arrayNesting(s)
    print("result is ${result}")
}

fun arrayNesting(nums: Array<Int>): Int {
    val visited = BooleanArray(nums.size)
    var res = 0
    for (i in nums.indices) {
        if (!visited[i]) {
            var start = nums[i]
            var count = 0
            do {
                start = nums[start]
                count++
                visited[start] = true
            } while (start != nums[i])
            res = Math.max(res, count)
        }
    }
    return res;

}

